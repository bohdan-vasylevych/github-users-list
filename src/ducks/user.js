import api from '../api';

import {takeLatest, put, call} from 'redux-saga/effects'

const LOAD_USERS_INFO = 'LOAD_USERS_INFO';
const LOAD_USERS_INFO_SUCCESS = 'LOAD_USERS_INFO_SUCCESS';
const LOAD_USERS_INFO_FAILURE = 'LOAD_FAILURE';

const initialState = {
    data: null,
    error: null,
    loading: false
};

export default function usersInfoReducer(state = initialState, action) {
    switch (action.type) {
        case LOAD_USERS_INFO:
            return {
                ...state,
                error: null,
                loading: true
            };
        case LOAD_USERS_INFO_SUCCESS:
            return {
                ...state,
                loading: false,
                data: action.data,
                error: null
            };
        case LOAD_USERS_INFO_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.error,
                data: null
            };
        default:
            return state;
    }
}

export function* fetchUsersInfo(userName) {
    const result = yield call(api.fetchUsersInfo, userName);
    if (result) {
        yield put({type: LOAD_USERS_INFO_SUCCESS, data: result});
    } else {
        yield put({type: LOAD_USERS_INFO_FAILURE, error: 'Something went wrong'});
    }
}
export function* watchRequest() {
    yield takeLatest(LOAD_USERS_INFO, fetchUsersInfo);
}

export function onUsersInfoFetch(userName) {
    return {
        type: LOAD_USERS_INFO,
        userName
    };
}
