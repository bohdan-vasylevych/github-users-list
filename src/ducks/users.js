import api from '../api';

import {takeLatest, put, call} from 'redux-saga/effects'

const LOAD_USERS = 'LOAD_USERS';
const LOAD_SUCCESS = 'LOAD_SUCCESS';
const LOAD_FAILURE = 'LOAD_FAILURE';
const LOAD_MORE_SUCCESS = 'LOAD_MORE_SUCCESS';
const LOAD_MORE = 'LOAD_MORE';

const initialState = {
    data: null,
    error: null,
    loading: false
};

export default function usersReducer(state = initialState, action) {
    switch (action.type) {
        case LOAD_USERS:
            return {
                ...state,
                error: null,
                loading: true
            };
        case LOAD_SUCCESS:
            return {
                ...state,
                loading: false,
                data: action.data,
                error: null
            };
        case LOAD_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.error,
                data: null
            };
        case LOAD_MORE:
            return {
                ...state,
                error: null,
                loading: true,
                id: null
            };
        case LOAD_MORE_SUCCESS:
            return {
                ...state,
                loading: false,
                data: [...state.data, ...action.data],
                error: null
            };
        default:
            return state;
    }
}

export function* fetchUsers() {
    const result = yield call(api.fetchUsers);
    if (result) {
        yield put({type: LOAD_SUCCESS, data: result})
    } else {
        yield put({type: LOAD_FAILURE, error: 'Something went wrong'})
    }
}

export function* fetchMoreUsers(id) {
    const result = yield call(api.fetchMoreUsers, id);
    if (result) {
        yield put({type: LOAD_MORE_SUCCESS, data: result})
    } else {
        yield put({type: LOAD_FAILURE, error: 'Something went wrong'})
    }
}

export function* watchRequest() {
    yield takeLatest(LOAD_USERS, fetchUsers);
    yield takeLatest(LOAD_MORE, fetchMoreUsers);
}

export function onUsersFetch() {
    return {
        type: LOAD_USERS
    };
}

export function onMoreUsersFetch(id) {
    return {
        type: LOAD_MORE,
        id
    };
}
