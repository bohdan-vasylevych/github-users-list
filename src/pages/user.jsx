import React, {Component} from 'react';
import {Grid, Row} from 'react-bootstrap';
import {connect} from 'react-redux';

import * as userDuck from '../ducks/user';
import UsersInfo from '../components/UsersInfo';

class UserPage extends Component {

    componentDidMount() {
        this.props.loadUserInfo(this.props.match.params.username)
    }

    render() {
        return (
            <Grid>
                <Row>
                    <UsersInfo {...this.props.data}/>
                </Row>
            </Grid>
        )
    }
}

export default connect(
    state => state.userInfo,
    dispatch => ({
        loadUserInfo: (userName) => dispatch(userDuck.onUsersInfoFetch(userName))
    })
)(UserPage);
