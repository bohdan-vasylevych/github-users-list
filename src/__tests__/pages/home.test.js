import React from 'react';
import {Provider} from 'react-redux'
import {applyMiddleware, combineReducers, createStore} from 'redux';
import createSagaMiddleware from 'redux-saga';
import rendered from 'react-test-renderer';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import HomePage from '../../pages/home';
import * as usersReducer from '../../ducks/users';
import * as usersInfoReducer from '../../ducks/user';

const sagaMiddleware = createSagaMiddleware();
const rootReducer = combineReducers({users: usersReducer.default, userInfo: usersInfoReducer.default});

const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
configure({adapter: new Adapter()});


describe('App component', () => {

    test('matches the snapshot', () => {
        const tree = rendered.create(<Provider store={store}><HomePage/></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });

});
