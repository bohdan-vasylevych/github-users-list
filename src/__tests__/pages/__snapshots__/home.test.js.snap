// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches the snapshot 1`] = `
<div
  className="home-page"
>
  <div
    className="user-list container"
  >
    <div
      className="row"
    />
  </div>
  <button
    className="btn btn-primary btn-block"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Load More
  </button>
</div>
`;
